packages/fast-usdc/src/exos/status-manager.js

```diff
@@ -1,12 +1,152 @@
+       * Queue a transaction before initiating a transfer. Receive
+       * the entry index in return.
+       *
+       * @param {CctpTxEvidence} evidence
+       * @returns {number}
+       */
+      observe(evidence) {
+        // XXX like `appendToStoredArray`, but returns index. consider
+        // upstreaming the change if it doesn't affect existing consumers
+        const key = getStatusKey(evidence);
+        const entry = { ...evidence, status: TxStatus.OBSERVED };
+        if (!txs.has(key)) {
+          // XXX append to store helper
+          txs.init(key, harden([entry]));
+          return 0;
+        }
+        const current = txs.get(key);
+        // XXX do we need defensive logic to ensure multiple entries don't have
+        // the same txHash+chainId?
+        txs.set(key, harden([...current, entry]));
+        return current.length;
+      },
+      /**
+       * Mark an `OBSERVED` transaction as `ADVANCED`.
+       *
+       * @param {string} address
+       * @param {bigint} amount
+       * @param {number} index
+       */
+      advance(address, amount, index) {
+        const key = `${address}-${amount}`;
+        const mutableEntries = [...txs.get(key)];
+        if (mutableEntries[index].status !== TxStatus.OBSERVED) {
+          throw makeError(
+            `Invalid status ${q(mutableEntries[index].status)} for ${q(key)}`,
+          );
+        }
+        mutableEntries[index] = {
+          ...mutableEntries[index],
+          status: TxStatus.ADVANCED,
+        };
+        txs.set(key, harden(mutableEntries));
+      },
+      /**
+       * Find an `ADVANCED` tx waiting to be `SETTLED`
+       *
+       * @param {string} address
+       * @param {bigint} amount
+       * @returns {boolean}
+       */
+      hasPendingSettlement(address, amount) {
+        const key = `${address}-${amount}`;
+        const entries = txs.get(key);
+        const unsettledIdx = entries.findIndex(
+          ({ status }) => status === TxStatus.ADVANCED,
+        );
+        // TODO: if OBSERVED -> SETTLED is a valid transition, so if the below
+        // is false we should probably also check for `OBSERVED`
+        return unsettledIdx > -1;
+      },
+      /**
+       * Mark an `ADVANCED` transaction as `SETTLED`.
+       *
+       * TODO - does this need to support `OBSERVED` -> `SETTLED`?
+       *
+       * @param {NobleAddress} address
+       * @param {bigint} amount
+       */
+      settle(address, amount) {
+        const key = `${address}-${amount}`;
+        const mutableEntries = [...txs.get(key)];
+        const unsettledIdx = mutableEntries.findIndex(
+          ({ status }) => status === TxStatus.ADVANCED,
+        );
+        if (unsettledIdx === -1) {
+          throw makeError(`No pending advance found for ${q(key)}`);
+        }
+        mutableEntries[unsettledIdx] = {
+          ...mutableEntries[unsettledIdx],
+          status: TxStatus.SETTLED,
+        };
+        txs.set(key, harden(mutableEntries));
```

Comment:
Do we ever delete from `txs`?

Reply:
Updated the approach to remove entries once they get to **Settled**